pkg/kv/kvserver/apply/asserter.go

```diff
@@ -0,0 +1,142 @@
+	if _, ok := a.appliedCmds[rangeID]; !ok {
+		a.appliedCmds[rangeID] = map[kvserverbase.CmdIDKey]*appliedCmd{}
+	}
+	if _, ok := a.appliedIndex[rangeID]; !ok {
+		a.appliedIndex[rangeID] = map[roachpb.ReplicaID]kvpb.RaftIndex{}
+	}
+}
+
+// Apply records a command being applied, and asserts validity.
+//
+// It assumes that entry application is durable, i.e. that nodes won't lose
+// applied state. This is not true for real CRDB nodes which can lose applied
+// state on restart, but is generally true in tests. If needed, we can loosen
+// this requirement by signalling truncation of applied state on node restart.
+func (a *Asserter) Apply(
+	rangeID roachpb.RangeID,
+	replicaID roachpb.ReplicaID,
+	cmdID kvserverbase.CmdIDKey,
+	entry raftpb.Entry,
+) {
+	if len(cmdID) == 0 {
+		return
+	}
+
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	a.ensureRangeLocked(rangeID)
+
+	// Assert and record the applied index.
+	if appliedIndex := a.appliedIndex[rangeID][replicaID]; kvpb.RaftIndex(entry.Index) <= appliedIndex {
+		panic(fmt.Sprintf("applied index regression for r%d/%d: %d -> %d", rangeID, replicaID, appliedIndex, entry.Index))
+	}
+	a.appliedIndex[rangeID][replicaID] = kvpb.RaftIndex(entry.Index)
+
+	if ac, ok := a.appliedCmds[rangeID][cmdID]; !ok {
+		// New command was applied, record it.
+		a.appliedCmds[rangeID][cmdID] = &appliedCmd{
+			cmdID:    cmdID,
+			rangeID:  rangeID,
+			logIndex: kvpb.RaftIndex(entry.Index),
+			replicas: map[roachpb.ReplicaID]bool{replicaID: true},
+		}
+	} else if ac.logIndex == kvpb.RaftIndex(entry.Index) {
+		// Command applying at the expected index, record the replica.
+		ac.replicas[replicaID] = true
+	} else {
+		// Applied command at unexpected log index, bail out.
+		var replicas []roachpb.ReplicaID
+		for id := range ac.replicas {
+			replicas = append(replicas, id)
+		}
+		msg := fmt.Sprintf("command %s re-applied at index %d on r%d/%d\n", cmdID, entry.Index, rangeID, replicaID)
+		msg += fmt.Sprintf("previously applied at index %d on replicas %s\n", ac.logIndex, replicas)
+		msg += fmt.Sprintf("entry: %+v", entry) // TODO: this formatting is not helpful
+		panic(msg)
+	}
+}
+
+// ApplySnapshot records snapshot application at the given applied index. The
+// replica is considered to have applied all commands up to and including the
+// applied index.
+func (a *Asserter) ApplySnapshot(
+	rangeID roachpb.RangeID,
+	replicaID roachpb.ReplicaID,
+	index kvpb.RaftIndex,
+) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	a.ensureRangeLocked(rangeID)
+
+	// Record the applied index.
+	a.appliedIndex[rangeID][replicaID] = index
+
+	// Record the applied commands for this replica. A command must have applied
+	// on at least one replica before it can be included in a snapshot, so all
+	// relevant commands are guaranteed to be present in the map.
+	for _, cmd := range a.appliedCmds[rangeID] {
+		if cmd.logIndex <= index {
+			cmd.replicas[replicaID] = true
+		} else {
+			delete(cmd.replicas, replicaID)
+		}
+	}
```

Comment:
Not following this section. When we apply a snapshot, there is no evidence that either of the commands has been applied, so shouldn't do `cmd.replicas[replicaID] = true`.

Also not following why `delete(cmd.replicas, replicaID)` for higher indices. Could you clarify? Does a snapshot "cancel" all previously applied commands? Should we allow that?

Reply:
I think we should allow this. Applying a snapshot is like applying many commands. It should always move the applied state forward.

If we allow applying snapshot at lower indices, this jeopardizes durability. Say, one index is applied at a majority of nodes. We send a snapshot that downgrades the applied index -> now some entries may be applied on less than a majority of replicas.

Reply:
> When we apply a snapshot, there is no evidence that either of the commands has been applied

That's what a snapshot means: a snapshot at index I contains the result of applying all transitions (commands) <= I. Specifically, we want to assert that we don't apply a log entry with a command that was already applied via snapshot.

> Does a snapshot "cancel" all previously applied commands?

Yes, if we apply a snapshot that regresses the applied index then it effectively un-applies those commands.

Reply:
> a snapshot at index I contains the result of applying all transitions (commands) <= I

This is correct, but there is no evidence that the concrete commands with the `CommandID`s that we track are the ones that were applied in this snapshot. We only have evidence for those commands that were reported through the `Apply` call.

For the same reason, when a snapshot skips over some indices, we return an "ambiguous error" to the client, because we don't know whether the commands were applied or overridden by some other commands with lower indices.

Reply:
> Specifically, we want to assert that we don't apply a log entry with a command that was already applied via snapshot.

This we should verify via LAI regression checks. We should never have a command applied at higher index than a snapshot, but lower LAI than the snapsot's.

Reply:
>If we allow applying snapshot at lower indices, this jeopardizes durability. Say, one index is applied at a majority of nodes. We send a snapshot that downgrades the applied index -> now some entries may be applied on less than a majority of replicas.

In principle, it's safe as long as a quorum has durable logs up to the commit index. Consider the case where the quorum is at commit index 10, and we have a lagging follower at applied index 2. We can send it two snapshots at index 5 and then index 3 (which regresses its applied index) without changing the safety properties of the cluster (it still has a quorum at commit index 10). If we lost either of the quorum nodes, we can catch the follower up using the logs of the remaining replicas (those entries have already undergone consensus and no longer need quorum).

I'm not sure about what happens in practice though. We only catch up followers whose log lags behind the leader's truncated log, and the leader won't truncate its log below what it's already applied, so by definition the snapshot must move the applied index forward. However, there may be races where multiple snapshots are sent to a follower (e.g. with delegated snapshots), and an earlier snapshot replaces a later snapshot.

I guess we could assert that snapshots don't regress the applied index and see if something trips it. But I don't think it's an invariant we can rely on.

>This is correct, but there is no evidence that the concrete commands with the CommandIDs that we track are the ones that were applied in this snapshot.

That's an assumption we'll have to make here, unless we want to track the actual state at each log index. I'm not going to spend time on that right now. The consistency checker will detect this, and we run a consistency check after every kvnemesis run.

Reply:
> Consider the case where the quorum is at commit index 10, and we have a lagging follower at applied index 2. We can send it two snapshots at index 5 and then index 3 (which regresses its applied index) without changing the safety properties of the cluster (it still has a quorum at commit index 10).

I mean that we can't apply a snapshot that downgrades the **applied** index. Surely there can be snapshots below the committed index, but applying them is fine because in theory we can keep the corresponding suffix of the committed log (I don't know if it's true in practice, or we just wipe the log entirely when we apply a snapshot).

The durability of an entry on a replica may exist in either of the two (or both) ways:
- The entry is in the log.
- The entry is in the snapshot.

Once an entry is durable on a replica in either of the two ways, we can't revert it because raft relies on it for quorum purposes.

> I guess we could assert that snapshots don't regress the applied index and see if something trips it. But I don't think it's an invariant we can rely on.

I think we should rely on it. When we regress the applied index and there is no corresponding entry in the log (which likely can happen because the log can be truncated up to the applied index), we lost durability of this entry on this node.

> That's an assumption we'll have to make here, unless we want to track the actual state at each log index.

I disagree. We don't even make this assumption in the prod code. We simply don't know which entries are behind the applied snapshot. In the test-only code we should make strictly fewer assumptions.

I think we should record only the evidence, and then check that the evidence is self-consistent.

Reply:
Yeah, we probably can, because etcd/raft already asserts it (unless I'm missing something):

https://github.com/etcd-io/raft/blob/ee0fe9da492888b55fe183cf1a42931ad551ec6b/raft.go#L718-L720

https://github.com/etcd-io/raft/blob/1df762940b8c309a27cfafb086d767c0c7e3f58f/log.go#L329-L331

>I disagree. We don't even make this assumption in the prod code. We simply don't know which entries are behind the applied snapshot. In the test-only code we should make strictly fewer assumptions.

What do you mean? The applied state at any given applied index must be identical across replicas. If a replica tells us that it successfully applied a command at index I (which we record here) then the snapshot must include that command to be correct. Sure, it would be nice to actually verify snapshots (which is expensive and I won't spend time on now), but that shouldn't stop us from asserting what we can assert -- that we don't re-apply a command that we've supposedly) already received via a snapshot. What do we have to gain from that?

Reply:
Sure, done.